lib/ContentTypeParser.js

```diff
@@ -1,71 +1,184 @@
+/* eslint-disable no-useless-return */
 'use strict'
 
 function ContentTypeParser () {
   this.customParsers = {}
-  this.parserList = []
+  this.customParsers['application/json'] = defaultJsonParser
+  this.parserList = ['application/json']
+  this.cachedParsers = {}
 }
 
-ContentTypeParser.prototype.add = function (contentType, fn) {
+ContentTypeParser.prototype.add = function (contentType, opts, parser) {
   if (typeof contentType !== 'string') throw new Error('The content type should be a string')
   if (contentType.length === 0) throw new Error('The content type cannot be an empty string')
-  if (typeof fn !== 'function') throw new Error('The content type handler should be a function')
+  if (typeof parser !== 'function') throw new Error('The content type handler should be a function')
 
   if (this.hasParser(contentType)) {
     throw new Error(`Content type parser '${contentType}' already present.`)
   }
+
+  parser.asString = opts.asString === true
+  parser.asBuffer = opts.asBuffer === true
+
   if (contentType === '*') {
     this.parserList.push('')
-    this.customParsers[''] = fn
+    this.customParsers[''] = parser
   } else {
-    this.parserList.unshift(contentType)
-    this.customParsers[contentType] = fn
+    if (contentType !== 'application/json') {
+      this.parserList.unshift(contentType)
+    }
+    this.customParsers[contentType] = parser
   }
 }
 
 ContentTypeParser.prototype.hasParser = function (contentType) {
+  if (contentType === 'application/json') return false
   return contentType in this.customParsers
 }
 
 ContentTypeParser.prototype.fastHasHeader = function (contentType) {
-  if (!contentType) return this.hasParser('')
-
   for (var i = 0; i < this.parserList.length; i++) {
     if (contentType.indexOf(this.parserList[i]) > -1) return true
   }
   return false
 }
 
-ContentTypeParser.prototype.getHandler = function (contentType) {
-  if (contentType === undefined) return this.customParsers['']
-
+ContentTypeParser.prototype.getParser = function (contentType) {
   for (var i = 0; i < this.parserList.length; i++) {
     if (contentType.indexOf(this.parserList[i]) > -1) {
-      return this.customParsers[this.parserList[i]]
+      var parser = this.customParsers[this.parserList[i]]
+      this.cachedParsers[contentType] = parser
```

Comment:
This can be a security issue

Reply:
Why?

Reply:
If I use `"application/json" + randomString()` as "content-type" header, that object grows a lot.

Reply:
Great!